Add support for more complex strokes

birktj opened the pull request:
This patch adds more complex stroking, specifically it adds:
- Line caps (butt, round, square)
- Line joins (miter, round, bevel)
- Dash patterns

A couple of questions that needs answering before this is ready:
- I couldn't find a way to compile the documentation so I don't really know if the extra text on the stroke parameter of `line` and `rect` looks ok.
- Since stroke can be a dictionary now, I had to modify the `Cast` impl of `Sides` in a way that results in not so good error messages. That is, you won't get any error about the valid fields of `Sides`, only it's internal type: in this case `stroke`.
- ~Currently there are some dash patterns that will give different results when rendering to image and pdf (dash arrays with only one element)~

Related to #260

laurmaedje commented:
Awesome, I'll take a closer look later, but I like it!

ollelogdahl commented:
Coolio! I am currently working on gradients/filling patterns on ALL paint objects. Stuff is starting to come together!

birktj commented:
Hmm, maybe I should rename `color` to `paint` then

laurmaedje commented on library/src/visualize/line.rs:
```diff
@@ -40,9 +40,32 @@ pub struct LineElem {
     ///   to `{1pt}`.
     /// - A stroke combined from color and thickness using the `+` operator as
     ///   in `{2pt + red}`.
+    /// - A stroke described by a dict with any of the following keys: 
```
```suggestion
    /// - A stroke described by a dictionary with any of the following keys: 
```

laurmaedje commented on library/src/visualize/shape.rs:
```diff
@@ -47,6 +47,20 @@ pub struct RectElem {
+    ///         - An array with elements that specify the lengths of dashes and gaps, alternating.
+    ///           Elements can also be the string `"dot"` for a length equal to the line thickness.
+    ///         - A dict with the keys `array`, same as the array above, and `phase`, the offset to
+    ///           the start of the first dash.
```
The text below should be adapted a bit because this is also a dictionary already

laurmaedje commented on src/export/pdf/page.rs:
```diff
@@ -267,7 +276,35 @@ impl PageContext<'_, '_> {
-            self.content.set_line_width(stroke.thickness.to_f32());
+            self.content.set_line_width(thickness.to_f32());
+            if self.state.stroke.as_ref().map(|s| &s.line_cap) != Some(line_cap) {
+                let cap = match line_cap {
```
no need for a binding here, the move directly in `set_line_cap`'s args is fine I think

laurmaedje commented on src/export/pdf/page.rs:
```diff
@@ -267,7 +276,35 @@ impl PageContext<'_, '_> {
+            }
+            if self.state.stroke.as_ref().map(|s| &s.dash_pattern) != Some(dash_pattern) {
+                if let Some(pattern) = dash_pattern {
+                    self.content.set_dash_pattern(
```
could you run cargo fmt?

laurmaedje commented on src/export/render.rs:
```diff
@@ -312,9 +312,42 @@ fn render_shape(
+            LineCap::Square => sk::LineCap::Square,
+        };
+        let line_join = match line_join {
+            LineJoin::Miter => sk::LineJoin::Miter,
```
maybe we want some From impl's for typst -> tiny-skia and typst -> pdf-writer types?

laurmaedje commented on src/export/render.rs:
```diff
@@ -312,9 +312,42 @@ fn render_shape(
+        };
+
+        let dash = dash_pattern.as_ref().and_then(|pattern| {
+            // tiny-skia only allows dash patterns with an even number of elements,
```
does this workaround produce the same output?

birktj replied:
Yes, tested with pdf.js on firefox and zathura (poppler)

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -1,19 +1,86 @@
 use super::*;
 
+/// A the line cap of a stroke
```
```suggestion
/// The line cap of a stroke.
```
same below

laurmaedje replied:
also, can you reorder these types from general to specific, with stroke being more general and linecap being more specific.

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -1,19 +1,86 @@
+/// A line dash pattern
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct DashPattern<T = Length, DT = DashLength<T>> {
+    /// The dash array.
+    pub dash_array: Vec<DT>,
+    /// The dash phase.
+    pub dash_phase: T,
```
```suggestion
    /// The dash array.
    pub array: Vec<DT>,
    /// The dash phase.
    pub phase: T,
```
maybe we can drop those as the whole thing is already a dash pattern?

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -59,15 +146,124 @@ impl<T: Debug> Debug for PartialStroke<T> {
+    "loosely dashed" => vec![Abs::pt(3.0).into(), Abs::pt(6.0).into()].into(),
+    "dashdotted" => vec![Abs::pt(3.0).into(), Abs::pt(2.0).into(), DashLength::LineWidth, Abs::pt(2.0).into()].into(),
+    "densely dashdotted" => vec![Abs::pt(3.0).into(), Abs::pt(1.0).into(), DashLength::LineWidth, Abs::pt(1.0).into()].into(),
+    "loosely dashdotted" => vec![Abs::pt(3.0).into(), Abs::pt(4.0).into(), DashLength::LineWidth, Abs::pt(4.0).into()].into(),
```
not a huge fan of these names, but if we keep them, they should use kebab-case, e.g. `"loosely-dash-dotted"`.

birktj replied:
I don't have strong opinions here other than that I think we should predefine a couple of useful dash patterns. These names seem to have some precedence though as they are also used in matplotlib: https://matplotlib.org/stable/gallery/lines_bars_and_markers/linestyles.html

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -59,15 +146,124 @@ impl<T: Debug> Debug for PartialStroke<T> {
+
+        let paint = take::<Paint>(&mut dict, "color")?;
+        let thickness = take::<Length>(&mut dict, "thickness")?;
+        let line_cap = take::<LineCap>(&mut dict, "line_cap")?;
+        let line_join = take::<LineJoin>(&mut dict, "line_join")?;
```
```suggestion
        let cap = take::<LineCap>(&mut dict, "cap")?;
        let join = take::<LineJoin>(&mut dict, "join")?;
```
maybe we can drop the `line_` prefix everywhere?

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -59,15 +146,124 @@ impl<T: Debug> Debug for PartialStroke<T> {
+        let line_cap = take::<LineCap>(&mut dict, "line_cap")?;
+        let line_join = take::<LineJoin>(&mut dict, "line_join")?;
+        let dash_pattern = take::<Option<DashPattern>>(&mut dict, "dash")?;
+        let miter_limit = take::<f64>(&mut dict, "miter_limit")?;
```
```suggestion
        let miter_limit = take::<f64>(&mut dict, "miter-limit")?;
```

birktj commented:
Thanks for your feedback! I will be away for a couple of days now, but when I get back I will go through your suggestions.

birktj commented:
I have incorporated your suggestions now and also updated the `Debug` impl for `PartialStroke`. I also rebased ontop of latest main in order to fix a failing test and update the stroke documentation for `polygon`.

ollelogdahl commented:
Hey, would you mind having a bit of a discussion over at discord? (my handle is Gaffa#4308). I have communicated with reiknh, and have partially implemented gradients for filling. I think line caps and line joins are great as a separate commit, but i think dash-patterns is a special case of a pattern applied _along_ a line. The issue is here https://github.com/typst/typst/issues/443, but I think most can be discussed over discord if you have the time.

Right now there is no implementation of patterns, but when that comes I think the dash-patterns will be made using that.

laurmaedje commented:
Did you two discuss on Discord and, if so, what was the result?

birktj commented:
Just a little bit. My understanding is that patterns _along_ paths is quite far away atm so I don't think there is too much point in waiting for this. I also think that anyways it is nice to utilize PDF's native support for dash patterns which would probably be more efficient (in document size terms). Finally I think that having simple support for dash patterns in the public API is important even if they are implemented in a different way, and  there should be nothing in the way of changing the underlying implementation in the future.

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -48,13 +87,219 @@ impl PartialStroke<Abs> {
+/// The line join of a stroke
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub enum LineJoin {
+    Miter,
```
Would it make sense to have the limit be part of this? Something like `Miter { limit: f64 }`. And on the Typst side it would be something like `#set line(stroke: (thickness: 1pt, join: (kind: "miter", limit: 4.0)))`. I'm unsure, it's getting a bit crazy nested at that point.

birktj replied:
I guess it could be an okay alternative, but I don't know if there is much point. It would introduce a little bit of extra complexity in both the docs and the implementation without so much gain, I don't think the `miter_limit` field in the stroke type is too much in the way.
 
A problem would also be how to set the `miter_limit` in a `set` rule without also setting the line join.

laurmaedje commented on src/geom/stroke.rs:
```diff
@@ -48,13 +87,219 @@ impl PartialStroke<Abs> {
+}
+
+cast_from_value! {
+    LineCap: "line cap",
```
you could also derive `Cast` for these, then you get this for free. only works for string-like enums though.

birktj replied:
That's pretty neat, I will do that
